File: lib/combination.go
var factorialMemo []int64

func factorial(n, m int) int64 {
	if n <= 1 {
		return 1
	}
	if factorialMemo[n] != 0 {
		return factorialMemo[n]
	}
	ret := int64(n) * factorial(n-1, m) % int64(m)
	factorialMemo[n] = ret
	return ret
}

func combination(n, k, m int) int64 {
	mm := int64(m)
	x := factorial(n, m) % mm
	y := factorial(k, m) % mm
	y = y * factorial(n-k, m) % mm
	return x * pow(y, m-2, m) % mm
}

func permutation(n, k, m int) int64 {
	return combination(n+k-1, n-1, m)
}

func pow(p, n, m int64) int64 {
	if n == 0 {
		return 1
	}
	if n&1 == 0 {
		r := pow(p, n/2, m)
		return r * r % int64(m)
	}
	return int64(p) * pow(p, n-1, m) % int64(m)
}

File: lib/combination.rs
#[allow(dead_code)]
struct Combination {
    n: usize,
    fact: Vec<i64>,
    inv: Vec<i64>,
    m: i64,
}

#[allow(dead_code)]
impl Combination {
    fn new(n: usize, m: i64) -> Self {
        let mut fact = vec![1; n + 1];
        for i in 1..=n {
            fact[i] = fact[i - 1] * i as i64 % m;
        }
        let mut inv = vec![0; n + 1];
        inv[n] = Self::pow(fact[n], m - 2, m);
        for i in (1..=n).rev() {
            inv[i - 1] = inv[i] * i as i64 % m;
        }

        Self { n, fact, inv, m }
    }

    fn permutation(&self, n: usize, k: usize) -> i64 {
        self.fact[n] * self.inv[n - k] % self.m
    }

    fn choose(&self, n: usize, k: usize) -> i64 {
        self.fact[n] * self.inv[k] % self.m * self.inv[n - k] % self.m
    }

    fn pow(x: i64, exp: i64, m: i64) -> i64 {
        if exp == 0 {
            return 1;
        }
        let r = Self::pow(x, exp / 2, m);
        if exp & 1 == 0 {
            r * r % m
        } else {
            r * r % m * x % m
        }
    }
}
